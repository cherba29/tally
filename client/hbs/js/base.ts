export interface TallyAccount {
  name: string;
  description: string;

  // Month when account was opened and possibly closed.
  openedOn: string;
  closedOn: string | null;

  // Is this an external/bookeeping account. External accounts are not
  // considered to be part of any owner.
  external: boolean;

  // Is this synthetic summary account.
  summary: boolean;

  // List of account owner ids.
  owners: string[];

  // Account type, for example 'CREDIT_CARD'.
  type: string;

  // Url to the account.
  url: string;

  // Physical address for the account.
  address: string;

  // Phone number for customer support.
  phone: string;

  // Username/password to use to login to the account.
  userName: string;
  password: string;

  // Real account number associated with this account.
  number: string;
}

export interface Balance {
  amount: number;
  date: Date;
  type: string;
}

export interface Transaction {
  isExpense: boolean;
  isIncome: boolean;
  toAccountName: string;
  balance: Balance;
  balanceFromStart: number;
  balanceFromEnd: number;
  description: string;
}

export interface Statement {
  inFlows?: number;
  outFlows?: number;
  income?: number;
  totalPayments?: number;
  totalTransfers?: number;
  isClosed?: boolean;
  addSub?: number;
  startBalance?: Balance;
  endBalance?: Balance;
  isCovered?: boolean;
  isProjectedCovered?: boolean;
  hasProjectedTransfer?: boolean;
  change?: number;
  percentChange?: number;
  unaccounted?: number;
  transactions?: Transaction[];
}

export interface SummaryStatement {
  inFlows: number;
  outFlows: number;
  income: number;
  totalPayments: number;
  totalTransfers: number;
  isClosed?: boolean;
  accounts: string[];
  addSub: number;
  startBalance: Balance;
  endBalance: Balance;
  isCovered?: boolean;
  isProjectedCovered?: boolean;
  hasProjectedTransfer?: boolean;
  change?: number;
  percentChange?: number;
  unaccounted?: number;
  transactions?: Transaction[];
}
