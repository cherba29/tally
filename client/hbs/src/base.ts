import {Balance} from '@tally/lib/core/balance';
import {GqlTransaction} from './gql_types';

export interface SummaryStatement {
  inFlows: number;
  outFlows: number;
  income: number;
  totalPayments: number;
  totalTransfers: number;
  isClosed?: boolean;
  accounts: string[];
  addSub: number;
  startBalance?: Balance;
  endBalance?: Balance;
  isCovered?: boolean;
  isProjectedCovered?: boolean;
  hasProjectedTransfer?: boolean;
  change?: number;
  percentChange?: number;
  unaccounted?: number;
  transactions?: GqlTransaction[];
}
