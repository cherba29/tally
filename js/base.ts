export interface TallyAccount {
  name: string;
  description: string;

  // Month when account was opened and possibly closed.
  openedOn: string;
  closedOn: string | null;

  // Is this an external/bookeeping account. External accounts are not
  // considered to be part of any owner.
  external: boolean;

  // Is this synthetic summary account.
  summary: boolean;

  // List of account owner ids.
  owners: string[];

  // Account type, for example 'CREDIT_CARD'.
  type: string;

  // Url to the account.
  url: string;

  // Physical address for the account.
  address: string;

  // Phone number for customer support.
  phone: string;

  // Username/password to use to login to the account.
  userName: string;
  password: string;

  // Real account number associated with this account.
  number: string;
}

export interface Balance {
  amount: number;
  type: string;
}

export interface Transaction {
  isExpense: boolean;
  isIncome: boolean;
  toAccountName: string;
  balance: Balance;
}

export interface Statement {
  isClosed: boolean;
  addSub: number;
  endBalance: Balance;
  isCovered: boolean;
  isProjectedCovered: boolean;
  hasProjectedTransfer: boolean;
  percentChange: number;
  unaccounted: number;
  transactions: Transaction[];
}

export interface SummaryStatement {
  isClosed: boolean;
  accounts: string[];
  addSub: number;
  endBalance: Balance;
  isCovered: boolean;
  isProjectedCovered: boolean;
  hasProjectedTransfer: boolean;
  percentChange: number;
  unaccounted: number;
  transactions: Transaction[];
}
