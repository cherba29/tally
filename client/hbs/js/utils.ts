import { GqlBudget } from '@backend/types';
import {TallyAccount, Statement, SummaryStatement} from './base';
import {Cell} from './cell';
import {Row} from './row';

/**
 *  Create map of owner->type->accountname, it will be rendered in this format.
 * @param accountNameToAccount - account name->account map
 */
function getOwnerTypeAccountMap(
    accountNameToAccount: {[accountType: string]: TallyAccount}):
        {[ownerType:string]: {[accountType:string]: string[]}} {
  const ownerTypeAccountNames: {
    [ownerType:string]: {[accountType:string]: string[]}} = {};

  for (const accountName of Object.keys(accountNameToAccount)) {
    const account: TallyAccount = accountNameToAccount[accountName]
    for (const owner of account.owners) {
      let typeAccounts: {[accountName:string]: string[]};
      if (owner in ownerTypeAccountNames) {
        typeAccounts = ownerTypeAccountNames[owner];
      } else {
        ownerTypeAccountNames[owner] = typeAccounts = {}
      }

      let accountNames: string[];
      if (account.type in typeAccounts) {
        accountNames = typeAccounts[account.type];
      } else {
        typeAccounts[account.type] = accountNames = [];
      }
      accountNames.push(account.name);
    }
  }
  return ownerTypeAccountNames;
}

interface PopupMonthSummaryData {
  id: string,
  accountName: string,
  month: string,
  summary?: SummaryStatement,
  statements?: StatementEntry[]
}

interface MonthTransferMatrix {
  accounts: {
    name: string;
    transfers: {[accountName:string]: number},
    total:number
  }[],
  total: number
}

interface StatementEntry {
  name: string,
  stmt: Statement
}

interface PopupMonthData {
  id: string,
  accountName: string,
  month: string,
  stmt: Statement
}

interface HeadingPopupData {
  id: string,
  account: TallyAccount
}

export type PopupData = PopupMonthSummaryData | PopupMonthData | HeadingPopupData;


interface CellData {
  cells: Cell[],
  popups: PopupData[],
}

// Computes data for summary cells.
function getSummaryCells(
    owner: string, name: string, months: string[],
    statements: {[accountName:string]: {[month:string]: Statement}},
    summaries: {[month:string]: SummaryStatement}): CellData {

  const cellData: CellData = {
    cells: [],
    popups: []
  };
  for (const month of months) {
    const summaryStmt: SummaryStatement = summaries[month];
    const id = owner + "_" + name + "_" + month;
    cellData.cells.push(new Cell(id, summaryStmt));
    const accounts = ('accounts' in summaryStmt) ? summaryStmt.accounts : [];
    cellData.popups.push({
      id,
      accountName: owner + " " + name,
      month,
      summary: summaryStmt,
      statements: accounts.map((accountName: string): StatementEntry => ({
          name: accountName, stmt: statements[accountName][month] }))
    });
  }
  return cellData;
}

function getPopupData(
  id: string, owner: string, name: string,
  month: string, stmt: Statement): PopupMonthData {
  return {
    id,
    accountName: name,
    month,
    stmt
  };
}

interface MatrixDataView {
  months: string[],
  rows: Row[],
  popupCells: PopupData[]
}

/**
 * Builds dataView structure with months, rows and popup cell data.
 *
 * @param {Array<string>} months - descending, continuous list of months to show
 * @param accountNameToAccount - account name to account map
 * @param statements - table indexed by account name and month of statements
 * @param summaries - table indexed by owner name + account type and month
 *                    of statement summaries.
 */
export function transformBudgetData(
    months: string[], accountNameToAccount: {[accountName:string]: TallyAccount},
    statements: {[accountName:string]: {[month:string]: Statement}},
    summaries: {[ownerAccountType:string]: {[month:string]: SummaryStatement}}) {
  const dataView: MatrixDataView = {
    months,
    rows: [],
    popupCells: []
  };
  const ownerTypeAccountNames = getOwnerTypeAccountMap(accountNameToAccount);
  const ownersSorted = getKeysSorted(ownerTypeAccountNames);

  for (const owner of ownersSorted) {
    const typeAccountNames = ownerTypeAccountNames[owner];
    dataView.rows.push(new Row(owner, 'SPACE', []));
    const summaryName = owner + " SUMMARY";
    if (summaryName in summaries) {
      const cellData = getSummaryCells(
          owner, "SUMMARY", months, statements, summaries[summaryName]);
      dataView.rows.push(new Row(owner, 'TOTAL', cellData.cells));
      Array.prototype.push.apply(dataView.popupCells, cellData.popups);
    }
    const accountTypes = getKeysSorted(typeAccountNames);

    for (const accountType of accountTypes) {
      dataView.rows.push(new Row(
          "*** " + owner + " - " + accountType + " *** accounts", 'SPACE', []));
      const accountNames = typeAccountNames[accountType].sort(
          (a, b) => a < b ? -1 : (a > b ? 1 : 0));
      for (const accountName of accountNames) {
        const accountStatements = statements[accountName];
        const cells = [];
        for (const month of months) {
          const stmt = accountStatements[month];
          const id = owner + "_" + accountName + "_" + month;
          cells.push(new Cell(id, stmt));
          dataView.popupCells.push(
              getPopupData(id, owner, accountName, month, stmt))
        }
        const account = accountNameToAccount[accountName];
        dataView.rows.push(new Row(account, 'NORMAL', cells));
        dataView.popupCells.push({
          id: account.name,
          account
        });
      }
      const name = owner + " " + accountType;
      const cellData = getSummaryCells(
        owner, accountType, months, statements, summaries[name]);
      dataView.rows.push(new Row(name, 'TOTAL', cellData.cells));
      Array.prototype.push.apply(dataView.popupCells, cellData.popups);
    }
  }
  return dataView;
}

/**
 * Builds dataView structure with months, rows and popup cell data.
 *
 * @param data - Server returnred GqlBudget structure.
 */
export function transformGqlBudgetData(data: GqlBudget) {
  const accountNameToAccount: {[accountName:string]: TallyAccount} = {};
  // Object.assign({}, ...data.accounts.map((x) => ({[x.name]: x})));;
  for (const account of data.accounts) {
    accountNameToAccount[account.name] = {
      name: account.name,
      description: account.description,
      openedOn: account.openedOn,
      closedOn: account.closedOn,
      owners: account.owners ?? [],
      url: account.url,
      type: account.type,
      address: account.address,
      userName: account.userName,
      number: account.number,
      phone: account.phone,
      password: account.password,
      summary: account.summary,
      external: account.external,
    };
  }
  // account name => month => statement map.
  const statements: {[accountName:string]: {[month:string]: Statement}} = {};
  for (const stmt of data.statements) {
    const entry = statements[stmt.name] || (statements[stmt.name] = {});
    if (stmt.isClosed) {
      entry[stmt.month] = {
        isClosed: true
      };
    } else {
      entry[stmt.month] = {
        inFlows: stmt.inFlows,
        outFlows: stmt.outFlows,
        isClosed: stmt.isClosed || undefined,
        isCovered: stmt.isCovered,
        isProjectedCovered: stmt.isProjectedCovered,
        hasProjectedTransfer: stmt.hasProjectedTransfer,
        income: stmt.income,
        totalPayments: stmt.totalPayments,
        totalTransfers: stmt.totalTransfers,
        ... (stmt.change == null ? {} : {change: stmt.change}),
        ... (stmt.percentChange == null ? {} : {percentChange: stmt.percentChange}),
        ... (stmt.unaccounted == null ? {} : {unaccounted: stmt.unaccounted}),
        startBalance: (stmt.startBalance?.date == null ? null : {
          amount: stmt.startBalance.amount,
          date: stmt.startBalance.date,
          type: stmt.startBalance.type,
        }),
        endBalance: (stmt.endBalance?.date == null ? null : {
          amount: stmt.endBalance.amount,
          date: stmt.endBalance.date,
          type: stmt.endBalance.type,
        }),
        addSub: stmt.inFlows + stmt.outFlows,
        transactions: stmt.transactions?.map(t=>({
          toAccountName: t.toAccountName,
          isExpense: t.isExpense,
          isIncome: t.isIncome,
          balance: {
            amount: t.balance.amount,
            date: t.balance.date,
            type: t.balance.type
          },
          ...(t.balanceFromStart != null && {balanceFromStart: t.balanceFromStart}),
          ...(t.balanceFromEnd != null && {balanceFromEnd: t.balanceFromEnd}),
          description: t.description || ''
        })) ?? [],
      };
    }
  }
  // owner + accout type => month => summary statement map.
  const summaries: {[ownerAccountType:string]: {[month:string]: SummaryStatement}} = {};
  for (const stmt of data.summaries) {
    const entry = summaries[stmt.name] || (summaries[stmt.name] = {});
    entry[stmt.month] = {
      accounts: stmt.accounts,
      inFlows: stmt.inFlows,
      outFlows: stmt.outFlows,
      income: stmt.income,
      totalPayments: stmt.totalPayments,
      totalTransfers: stmt.totalTransfers,
      addSub: stmt.addSub,
      startBalance: stmt.startBalance && {
        amount: stmt.startBalance.amount,
        date: stmt.startBalance.date,
        type: stmt.startBalance.type,
      },
      endBalance: stmt.endBalance && {
        amount: stmt.endBalance.amount,
        date: stmt.endBalance.date,
        type: stmt.endBalance.type,
      },
      ... (stmt.change == null ? {} : {change: stmt.change}),
      ... (stmt.percentChange == null ? {} : {percentChange: stmt.percentChange}),
      ... (stmt.unaccounted == null ? {} : {unaccounted: stmt.unaccounted}),
    };
  }
  return transformBudgetData(data.months, accountNameToAccount, statements, summaries);

  // return {
  //   months: data.months,
  //   accountNameToAccount,
  //   statements,
  //   summaries
  // };
}

function getKeysSorted(obj:  {[key:string]: any;}): string[] {
  return Object.keys(obj).sort((a, b) => a < b ? -1 : (a > b ? 1 : 0));
}
