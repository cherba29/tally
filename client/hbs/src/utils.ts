import {GqlAccount, GqlStatement, GqlSummaryStatement, GqlTableRow} from './gql_types';

export interface PopupMonthSummaryData {
  accountName: string;
  month: string;
  summary?: GqlSummaryStatement;
  statements?: StatementEntry[];
}

export interface StatementEntry {
  name: string;
  stmt: GqlStatement;
}

export interface PopupMonthData {
  accountName: string;
  month: string;
  stmt: GqlStatement;
}

export interface HeadingPopupData {
  account: GqlAccount;
}

export type PopupData = PopupMonthSummaryData | PopupMonthData | HeadingPopupData;

export type Rows = {
  [owner: string]: GqlTableRow[];
};
