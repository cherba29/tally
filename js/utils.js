"use strict";

// Create map of owner->type->accountname, it will be rendered in this format.
// @param accountNameToAccount - account name->account map
function getOwnerTypeAccountMap(accountNameToAccount) {
  console.log("name to account", accountNameToAccount);
  var ownerTypeAccountNames = {};

  for (var accountName in accountNameToAccount) {
    var account = accountNameToAccount[accountName]
    for (var i = 0; i < account.owners.length; i++) {
      var owner = account.owners[i];
      var typeAccounts;
      if (owner in ownerTypeAccountNames) {
        typeAccounts = ownerTypeAccountNames[owner];
      } else {
        ownerTypeAccountNames[owner] = typeAccounts = {}
      }

      var accountNames;
      if (account.type in typeAccounts) {
        accountNames = typeAccounts[account.type];
      } else {
        typeAccounts[account.type] = accountNames = [];
      }
      accountNames.push(account.name);
    }
  }
  return ownerTypeAccountNames;
}

// Data for rendering given row.
function Row(title, type, cells) {
  this.title = title;
  this.isSpace = ('SPACE' == type);
  this.isTotal = ('TOTAL' == type);
  this.isNormal = ('NORMAL' == type);
  this.cells = cells
}

// Data for rendering given cell.
function Cell(id, stmt) {
  this.isClosed = stmt.isClosed;
  if (!this.isClosed) {
    this.id = id;
    this.addSub = ('addSub' in stmt) ? stmt.addSub : null;
    if ('endBalance' in stmt && stmt.endBalance !== null) {
      this.balance = stmt.endBalance.amount;
      this.isProjected = stmt.endBalance.type != 'CONFIRMED';
    } else {
      this.balance = null;
      this.isProjected = false;
    }
    this.isCovered = stmt.isCovered;
    this.isProjectedCovered = stmt.isProjectedCovered;
    this.hasProjectedTransfer = stmt.hasProjectedTransfer;
    this.isProjected = this.isProjected || this.hasProjectedTransfer;
    this.percentChange = ('percentChange' in stmt) ? stmt.percentChange : null;
    if ('unaccounted' in stmt) {
      this.unaccounted = stmt.unaccounted;
      this.balanced = !this.unaccounted;
    } else {
      this.unaccounted = null;
      this.balanced = true;
    }
  }
}

// Computes data for summary cells.
function getSummaryCells(owner, name, months, statements, summaries) {
  var cellData = {
    cells: [],
    popups: []
  };
  for (var m = 0, mm = months.length; m < mm; m++) {
    var month = months[m];
    var summaryStmt = summaries[month];
    var id = owner + "_" + name + "_" + month;
    cellData.cells.push(new Cell(id, summaryStmt));
    var accounts = ('accounts' in summaryStmt) ? summaryStmt.accounts : [];
    var transfer_matrix = buildMonthTransferMatrix(month, statements);
    console.log("transfer matrix", id, transfer_matrix);
    cellData.popups.push({
      id: id,
      accountName: owner + " " + name,
      month: month,
      summary: summaryStmt,
      transfer_matrix: transfer_matrix,
      statements: accounts.map(
          function(name) { return { name: name, stmt: statements[name][month] };})
    });
  }
  return cellData;
}

function getPopupData(id, owner, name, month, stmt) {
  return {
    id: id,
    accountName: name,
    month: month,
    stmt: stmt
  };
}

// Builds a list of accounts with name, map of account name to
// total transfers to that account, and total transfers for the whole account.
// @param statements - table indexed by account name and month of statements
function buildMonthTransferMatrix(month, statements) {
  var accounts = [];
  var overall_total = 0.0;
  //console.log("transfer matrix month ", month);
  for (var accountName in statements) {
    //console.log("transfer matrix accountName ", accountName);
    var stmt = statements[accountName][month];
    //console.log("transfer matrix stmt ", stmt);
    if (stmt.isClosed) {
      continue;
    }
    var transfers = [];
    var total = 0.0;
    for (var i = 0, ii = stmt.transactions.length; i < ii; i++) {
      var transaction = stmt.transactions[i];
      if (!transaction.isExpense && !transaction.isIncome) {
        transfers[transaction.toAccountName] = transaction.balance.amount;
        total += transaction.balance.amount
      }
    }
    overall_total += total;
    accounts.push({
      name: accountName,
      transfers: transfers,
      total: total
    });
  }
  accounts.sort(function(a, b) {
    return a.name < b.name ? -1 : (a.name > b.name ? 1 : 0)
  });
  return {
    accounts: accounts,
    total: overall_total
  };
}

// Builds dataView structure with months, rows and popup cell data.
// @param months - descending, continuous list of months to show
// @param accountNameToAccount - account name to account map
// @param statements - table indexed by account name and month of statements
// @param summaries - table indexed by owner name + account type and month
//                    of statement summaries.
function transformBudgetData(
    months, accountNameToAccount, statements, summaries) {
  var dataView = {
    months: months,
    rows: [],
    popupCells: []
  };
  var ownerTypeAccountNames = getOwnerTypeAccountMap(accountNameToAccount);
  console.log("ownerTypeAccountNames", ownerTypeAccountNames);
  var ownersSorted = getKeysSorted(ownerTypeAccountNames);
  console.log("ownersSorted", ownersSorted);
  for (var i = 0, ii = ownersSorted.length; i < ii; i++) {
    var owner = ownersSorted[i];
    var typeAccountNames = ownerTypeAccountNames[owner];
    dataView.rows.push(new Row(owner, 'SPACE', []));
    var summaryName = owner + " SUMMARY";
    //console.log("summaryName", summaryName);
    if (summaryName in summaries) {
      var cellData = getSummaryCells(
          owner, "SUMMARY", months, statements, summaries[summaryName]);
      dataView.rows.push(new Row(owner, 'TOTAL', cellData.cells));
      Array.prototype.push.apply(dataView.popupCells, cellData.popups);
    }
    var accountTypes = getKeysSorted(typeAccountNames);

    for (var j = 0, jj = accountTypes.length; j < jj; j++) {
      var accountType = accountTypes[j];
      dataView.rows.push(new Row(
          "*** " + owner + " - " + accountType + " *** accounts", 'SPACE', []));
      var accountNames = typeAccountNames[accountType].sort(function(a, b) {
        return a < b ? -1 : (a > b ? 1 : 0)
      });
      for (var k = 0, kk = accountNames.length; k < kk; k++) {
        var accountName = accountNames[k];
        var accountStatements = statements[accountName];
        var cells = [];
        for (var m = 0, mm = months.length; m < mm; m++) {
          var month = months[m];
          var stmt = accountStatements[month];
          var id = owner + "_" + accountName + "_" + month;
          //console.log("stmt", stmt, "id", id);
          cells.push(new Cell(id, stmt));
          dataView.popupCells.push(getPopupData(id, owner, accountName, month, stmt))
        }
        var account = accountNameToAccount[accountName];
        dataView.rows.push(new Row(account, 'NORMAL', cells));
        dataView.popupCells.push({
          id: account.name,
          account: account
        });
      }
      var name = owner + " " + accountType;
      var cellData = getSummaryCells(
        owner, accountType, months, statements, summaries[name]);
      dataView.rows.push(new Row(name, 'TOTAL', cellData.cells));
      Array.prototype.push.apply(dataView.popupCells, cellData.popups);
    }
  }
  console.log("dataView", dataView);
  return dataView;
}

function getKeysSorted(obj) {
  return Object.keys(obj).sort(function(a, b) {
    return a < b ? -1 : (a > b ? 1 : 0)
  });
}
