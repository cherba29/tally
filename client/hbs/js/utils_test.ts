import { Statement, SummaryStatement, TallyAccount } from './base';
import { transformBudgetData } from './utils';

describe("transformBudgetData", function () {

  it("works on empty", function () {
    const months: string[] = [];
    const accountNameToAccount = {};
    const statements = {};
    const summaries = {};
    const data_view = transformBudgetData(
        months, accountNameToAccount, statements, summaries);

    expect(data_view).toEqual({
      months: [],
      rows: [],
      popupCells: []
    });
  });

  it("works on single", function () {
    const months = ["Sep2014"];
    const account: TallyAccount = {
      name: "main",
      address: "some place",
      closedOn: null,
      description: "Main account",
      external: false,
      number: "123",
      openedOn: "Aug2014",
      owners: ["owner1", "owner2"],
      password: "password",
      phone: "123-456-7890",
      summary: false,
      type: "credit",
      url: "http://somewhere",
      userName: "user1"
    };
    const accountNameToAccount: {[accountName:string]: TallyAccount} = { main: account };
    const statement: Statement =  {
      isClosed: true,
      addSub: 100,
      inFlows: 150,
      outFlows: 50,
      income: 0,
      totalPayments: 0,
      totalTransfers: 0,
      startBalance: {
        amount: 2900,
        date: new Date('2020-04-03'),
        type: 'PROJECTED',
      },
      endBalance: {
        amount: 3000,
        date: new Date('2020-04-30'),
        type: 'PROJECTED',
      },
      isCovered: true,
      isProjectedCovered: true,
      hasProjectedTransfer: true,
      percentChange: 1.4,
      unaccounted: 500,
      transactions: []
    };
    const statements: {[accountName:string]: {[month:string]: Statement}} = {
      main: { Sep2014: statement }
    };
    const summaries: {[ownerAccountType:string]: {[month:string]: SummaryStatement}} = {
      'owner1 credit': {
        Sep2014: {
          isClosed: true,
          accounts: ["main"],
          addSub: 100,
          inFlows: 150,
          outFlows: 50,
          income: 0,
          totalPayments: 0,
          totalTransfers: 0,
          startBalance: {
            amount: 2900,
            date: new Date('2020-04-03'),
            type: 'PROJECTED',
          },
          endBalance: {
            amount: 3000,
            date: new Date('2020-04-30'),
            type: 'PROJECTED',
          },
          isCovered: true,
          isProjectedCovered: true,
          hasProjectedTransfer: true,
          percentChange: 1.4,
          unaccounted: 200,
          transactions: []
        }
      },
      'owner2 credit': {
        Sep2014: {
          isClosed: true,
          accounts: ["main"],
          addSub: 100,
          inFlows: 150,
          outFlows: 50,
          income: 0,
          totalPayments: 0,
          totalTransfers: 0,
          startBalance: {
            amount: 2900,
            date: new Date('2020-04-30'),
            type: 'PROJECTED',
          },
          endBalance: {
            amount: 3000,
            date: new Date('2020-04-03'),
            type: 'PROJECTED',
          },
          isCovered: true,
          isProjectedCovered: true,
          hasProjectedTransfer: true,
          percentChange: 1.4,
          unaccounted: 200,
          transactions: []
        }
      }
    };
    const data_view = transformBudgetData(
        months, accountNameToAccount, statements, summaries);

    expect(data_view).toMatchSnapshot();
  });

});
