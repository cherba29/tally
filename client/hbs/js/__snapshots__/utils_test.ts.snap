// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformBudgetData works on single 1`] = `
Object {
  "months": Array [
    "Sep2014",
  ],
  "popupCells": Array [
    Object {
      "accountName": "main",
      "id": "owner1_main_Sep2014",
      "month": "Sep2014",
      "stmt": Object {
        "addSub": 100,
        "endBalance": Object {
          "amount": 3000,
          "date": 2020-04-30T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "hasProjectedTransfer": true,
        "inFlows": 150,
        "income": 0,
        "isClosed": true,
        "isCovered": true,
        "isProjectedCovered": true,
        "outFlows": 50,
        "percentChange": 1.4,
        "startBalance": Object {
          "amount": 2900,
          "date": 2020-04-03T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "totalPayments": 0,
        "totalTransfers": 0,
        "transactions": Array [],
        "unaccounted": 500,
      },
    },
    Object {
      "account": Object {
        "address": "some place",
        "closedOn": null,
        "description": "Main account",
        "external": false,
        "name": "main",
        "number": "123",
        "openedOn": "Aug2014",
        "owners": Array [
          "owner1",
          "owner2",
        ],
        "password": "password",
        "phone": "123-456-7890",
        "summary": false,
        "type": "credit",
        "url": "http://somewhere",
        "userName": "user1",
      },
      "id": "main",
    },
    Object {
      "accountName": "owner1 credit",
      "id": "owner1_credit_Sep2014",
      "month": "Sep2014",
      "statements": Array [
        Object {
          "name": "main",
          "stmt": Object {
            "addSub": 100,
            "endBalance": Object {
              "amount": 3000,
              "date": 2020-04-30T00:00:00.000Z,
              "type": "PROJECTED",
            },
            "hasProjectedTransfer": true,
            "inFlows": 150,
            "income": 0,
            "isClosed": true,
            "isCovered": true,
            "isProjectedCovered": true,
            "outFlows": 50,
            "percentChange": 1.4,
            "startBalance": Object {
              "amount": 2900,
              "date": 2020-04-03T00:00:00.000Z,
              "type": "PROJECTED",
            },
            "totalPayments": 0,
            "totalTransfers": 0,
            "transactions": Array [],
            "unaccounted": 500,
          },
        },
      ],
      "summary": Object {
        "accounts": Array [
          "main",
        ],
        "addSub": 100,
        "endBalance": Object {
          "amount": 3000,
          "date": 2020-04-30T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "hasProjectedTransfer": true,
        "inFlows": 150,
        "income": 0,
        "isClosed": true,
        "isCovered": true,
        "isProjectedCovered": true,
        "outFlows": 50,
        "percentChange": 1.4,
        "startBalance": Object {
          "amount": 2900,
          "date": 2020-04-03T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "totalPayments": 0,
        "totalTransfers": 0,
        "transactions": Array [],
        "unaccounted": 200,
      },
    },
    Object {
      "accountName": "main",
      "id": "owner2_main_Sep2014",
      "month": "Sep2014",
      "stmt": Object {
        "addSub": 100,
        "endBalance": Object {
          "amount": 3000,
          "date": 2020-04-30T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "hasProjectedTransfer": true,
        "inFlows": 150,
        "income": 0,
        "isClosed": true,
        "isCovered": true,
        "isProjectedCovered": true,
        "outFlows": 50,
        "percentChange": 1.4,
        "startBalance": Object {
          "amount": 2900,
          "date": 2020-04-03T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "totalPayments": 0,
        "totalTransfers": 0,
        "transactions": Array [],
        "unaccounted": 500,
      },
    },
    Object {
      "account": Object {
        "address": "some place",
        "closedOn": null,
        "description": "Main account",
        "external": false,
        "name": "main",
        "number": "123",
        "openedOn": "Aug2014",
        "owners": Array [
          "owner1",
          "owner2",
        ],
        "password": "password",
        "phone": "123-456-7890",
        "summary": false,
        "type": "credit",
        "url": "http://somewhere",
        "userName": "user1",
      },
      "id": "main",
    },
    Object {
      "accountName": "owner2 credit",
      "id": "owner2_credit_Sep2014",
      "month": "Sep2014",
      "statements": Array [
        Object {
          "name": "main",
          "stmt": Object {
            "addSub": 100,
            "endBalance": Object {
              "amount": 3000,
              "date": 2020-04-30T00:00:00.000Z,
              "type": "PROJECTED",
            },
            "hasProjectedTransfer": true,
            "inFlows": 150,
            "income": 0,
            "isClosed": true,
            "isCovered": true,
            "isProjectedCovered": true,
            "outFlows": 50,
            "percentChange": 1.4,
            "startBalance": Object {
              "amount": 2900,
              "date": 2020-04-03T00:00:00.000Z,
              "type": "PROJECTED",
            },
            "totalPayments": 0,
            "totalTransfers": 0,
            "transactions": Array [],
            "unaccounted": 500,
          },
        },
      ],
      "summary": Object {
        "accounts": Array [
          "main",
        ],
        "addSub": 100,
        "endBalance": Object {
          "amount": 3000,
          "date": 2020-04-03T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "hasProjectedTransfer": true,
        "inFlows": 150,
        "income": 0,
        "isClosed": true,
        "isCovered": true,
        "isProjectedCovered": true,
        "outFlows": 50,
        "percentChange": 1.4,
        "startBalance": Object {
          "amount": 2900,
          "date": 2020-04-30T00:00:00.000Z,
          "type": "PROJECTED",
        },
        "totalPayments": 0,
        "totalTransfers": 0,
        "transactions": Array [],
        "unaccounted": 200,
      },
    },
  ],
  "rows": Array [
    Row {
      "cells": Array [],
      "isNormal": false,
      "isSpace": true,
      "isTotal": false,
      "title": "owner1",
    },
    Row {
      "cells": Array [],
      "isNormal": false,
      "isSpace": true,
      "isTotal": false,
      "title": "*** owner1 - credit *** accounts",
    },
    Row {
      "cells": Array [
        Cell {
          "addSub": 100,
          "balance": 3000,
          "balanced": false,
          "hasProjectedTransfer": true,
          "id": "owner1_main_Sep2014",
          "isClosed": true,
          "isCovered": true,
          "isProjected": true,
          "isProjectedCovered": true,
          "percentChange": 1.4,
          "unaccounted": 500,
        },
      ],
      "isNormal": true,
      "isSpace": false,
      "isTotal": false,
      "title": Object {
        "address": "some place",
        "closedOn": null,
        "description": "Main account",
        "external": false,
        "name": "main",
        "number": "123",
        "openedOn": "Aug2014",
        "owners": Array [
          "owner1",
          "owner2",
        ],
        "password": "password",
        "phone": "123-456-7890",
        "summary": false,
        "type": "credit",
        "url": "http://somewhere",
        "userName": "user1",
      },
    },
    Row {
      "cells": Array [
        Cell {
          "addSub": 100,
          "balance": 3000,
          "balanced": false,
          "hasProjectedTransfer": true,
          "id": "owner1_credit_Sep2014",
          "isClosed": true,
          "isCovered": true,
          "isProjected": true,
          "isProjectedCovered": true,
          "percentChange": 1.4,
          "unaccounted": 200,
        },
      ],
      "isNormal": false,
      "isSpace": false,
      "isTotal": true,
      "title": "owner1 credit",
    },
    Row {
      "cells": Array [],
      "isNormal": false,
      "isSpace": true,
      "isTotal": false,
      "title": "owner2",
    },
    Row {
      "cells": Array [],
      "isNormal": false,
      "isSpace": true,
      "isTotal": false,
      "title": "*** owner2 - credit *** accounts",
    },
    Row {
      "cells": Array [
        Cell {
          "addSub": 100,
          "balance": 3000,
          "balanced": false,
          "hasProjectedTransfer": true,
          "id": "owner2_main_Sep2014",
          "isClosed": true,
          "isCovered": true,
          "isProjected": true,
          "isProjectedCovered": true,
          "percentChange": 1.4,
          "unaccounted": 500,
        },
      ],
      "isNormal": true,
      "isSpace": false,
      "isTotal": false,
      "title": Object {
        "address": "some place",
        "closedOn": null,
        "description": "Main account",
        "external": false,
        "name": "main",
        "number": "123",
        "openedOn": "Aug2014",
        "owners": Array [
          "owner1",
          "owner2",
        ],
        "password": "password",
        "phone": "123-456-7890",
        "summary": false,
        "type": "credit",
        "url": "http://somewhere",
        "userName": "user1",
      },
    },
    Row {
      "cells": Array [
        Cell {
          "addSub": 100,
          "balance": 3000,
          "balanced": false,
          "hasProjectedTransfer": true,
          "id": "owner2_credit_Sep2014",
          "isClosed": true,
          "isCovered": true,
          "isProjected": true,
          "isProjectedCovered": true,
          "percentChange": 1.4,
          "unaccounted": 200,
        },
      ],
      "isNormal": false,
      "isSpace": false,
      "isTotal": true,
      "title": "owner2 credit",
    },
  ],
}
`;
